rfd/0002-custom-terraform-registry.md

```diff
@@ -118,33 +314,60 @@ provider consists of:
-the Terraform provider.
-
-By analogy with the main Teleport release system, the registry-compatible
-provider package will be created during release promotion, with a build task
-that will:
-
-1. For each release tarball (one each for multiple release platforms):
-    1. repack the tarball into an appropriately formatted & named zipfile
-    2. generate the checksum file from the zipfile
-    3. generate the signature file by signing the checksum file
-2. Upload all generated files to the AWS bucket that backs the registry
-3. Update the registry index to include the newly released providers (see
-   _Populating the registry_, below)
-
-I expect this to be implemented as a small Go program, rather than a shell
-script. Using small Go programs for build tasks has shown itself to be a
-useful, flexible and above all _legible_ tool for scripting build tasks in
-the mainline Teleport repository, and we should continue that pattern in
-`teleport-plugins`.
+For each release tarball (one each for multiple release platforms - i.e.
+`linux` and `darwin`):
+
+1. deduce the version, OS and architecture of the release from the
+    release tarball filename
+2. repack the tarball into an appropriately formatted & named zipfile
+3. generate the checksum file from the zipfile
+4. generate the signature file by signing the checksum file
+
+#### Updating the registry
+
+To update the registry, we take the artefacts and data generated in the
+previous step and
+
+1. Download the existing `versions` index file (or create fresh, if
+    not present)
+2. Add entries for the new providers in to the index
+3. Create the `download` entries for each file
+
+I expect this process to be implemented a sa small Go program (or a small suite
+of them), rather than a shell script. Using small Go programs for build tasks
+has shown itself to be a useful, flexible and above all _legible_ tool for
+scripting build tasks in the mainline Teleport repository, and we should
+continue that pattern in `teleport-plugins`.
+
+#### Final publishing
+
+Upload the repackaged `zip` files, the updated `versions` index and the new
+`download` files to the S3 bucket.
+
+#### Synchronization Hazard
+
+The Download-Update-Publish cycle for updating the registry index is vulnerable
+to corruption if multiple independent updates occur simultaneously.
+
+Unfortunately, AWS S3 does not provide any built-in methods for preventing this
+sort of issue.
+
+It also apparently lacks the primitives that would allow us to layer one over S3:
+
+* There is no way to push an object to a bucket only if an object with the same
+  key does not already exist, which means we can't really implement a race free
+  lockfile-like protocol
+* There are no compare-and-swap style put primitives
+* While versioning _is_ supported, there is no way to say "only overwrite object
+  _X_ if its current version is _Y_".
+
+Our initial approach may have to be relying on the build serialisation tools
+provided by Drone.
```

Comment:
This is fine.  We'll get to chubby at some point in the future :laugh:.  AFAIK we haven't had any races during promotion yet.  They've all been earlier in the process.